File: src/day2/day2.clj
(ns day2.day2
  (:require [utils]
            [clojure.string :as str]))

;; part 1

(def cubes {"red" 12 "green" 13 "blue" 14})

(defn valid-cube? [cube]
    (let [[n c] (str/split cube #" ")]
      (>= (get cubes c) (utils/parse-int n))))

(defn valid-set? [set]
    (every? valid-cube? (str/split set #", ")))

(defn part-1 []
  (->> (utils/read-input)
       str/split-lines
       (map #(-> % (str/split #": ") second)) 
       (keep-indexed
         (fn [idx sets]
           (when (every? valid-set? (str/split sets #"; "))
             (inc idx))))
       (apply +)))

;; part 2

(defn parse-draws [sets]
  (reduce
    (fn [a [_ n c]]
      (update a c max (utils/parse-int n)))
    {"red" 0 "green" 0 "blue" 0}
    sets))

(defn part-2 []
  (->> (utils/read-input)
       str/split-lines
       (map #(-> % (str/split #": ") second))
       (map (comp (partial apply *) 
                  vals 
                  parse-draws 
                  (partial re-seq #"(\d+) (red|green|blue)")))
       (apply +)))

(comment
  (part-1)
  
  (part-2))

File: day2/day2.go
package main

import (
    _ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed input.txt
var input string

var cubes = map[string]int{
    "red": 12,
    "green": 13,
    "blue": 14,
}

func checkSets(sets string) bool {
    for _, set := range strings.Split(sets, "; ") {
        for _, cube := range strings.Split(set, ", ") {
            x := strings.Split(cube, " ")
            n, _ := strconv.Atoi(x[0])
            if cubes[x[1]] < n {
                return false
            }
        }
    }

    return true
}

func part1() {
    sum := 0;
    for i, game := range strings.Split(strings.TrimSpace(input), "\n") {
        sets := strings.Split(game, ": ")[1]
        if !checkSets(sets) {
            continue
        }

        sum += i + 1
   }

   fmt.Println(sum)
}

func part2() {
    sum := 0;
    for _, game := range strings.Split(strings.TrimSpace(input), "\n") {
        sets := strings.Split(game, ": ")[1]
        cubes := map[string]int{}
        for _, set := range strings.Split(sets, "; ") {
            for _, cube := range strings.Split(set, ", ") {
                x := strings.Split(cube, " ")
                n, _ := strconv.Atoi(x[0])
                cubes[x[1]] = max(cubes[x[1]], n)
            }
        }

        sum += cubes["red"] * cubes["green"] * cubes["blue"]
   }

   fmt.Println(sum)
}

func main() {
    part1()
    part2()
}
